compat/mingw.c

```diff
@@ -426,6 +426,67 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+static BOOL exists(const wchar_t *wpathname)
+{
+	HANDLE hnd = CreateFileW(wpathname, GENERIC_READ,
+				 FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
+				 OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+	if (hnd == INVALID_HANDLE_VALUE)
+		return FALSE;
+	CloseHandle(hnd);
+	return TRUE;
+}
+
+static BOOL is_exclusively_writable(const wchar_t *wpathname)
+{
+	HANDLE hnd = CreateFileW(wpathname, GENERIC_WRITE, 0, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
```

Comment:
What state does this function check for? If I understand correctly, if a file is opened with `FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE`, then we do not need to rename, so this check should be false in this case. So is the check only true if we mmap the file, i.e. when we cannot use `unlink`?

By the way, it would be very helpful to have a test which demonstrates the issue. Currently we only have tests without the `FILE_SHARE` flags, and we do not actually care for those.

Reply:
If the file was opened before with SHARE_WRITE then another process can only open it with SHARE_WRITE. Here I open it with 0 (no share), which only succeeds if the handle is exclusive.